bundles/org.openhab.ui.basic/src/main/java/org/openhab/ui/basic/internal/render/InputRenderer.java

```diff
@@ -58,38 +69,127 @@ public boolean canRender(Widget w) {
-            snippet = snippet.replace("%undef_state%", "");
-            snippet = snippet.replace("%data_state%", dataState);
-        }
-
         Item item = null;
         try {
             item = itemUIRegistry.getItem(w.getItem());
         } catch (ItemNotFoundException e) {
             logger.debug("Failed to retrieve item during widget rendering: {}", e.getMessage());
         }
         String dataType;
-        if (item != null && item.getAcceptedDataTypes().stream().anyMatch(o -> Number.class.isAssignableFrom(o))) {
-            dataType = "Number";
+        if (item != null && item.getAcceptedCommandTypes().stream()
+                .anyMatch(o -> (!StringType.class.isAssignableFrom(o) && Number.class.isAssignableFrom(o)))) {
+            dataType = "number";
+        } else if (item != null && item.getAcceptedCommandTypes().stream()
+                .anyMatch(o -> (!StringType.class.isAssignableFrom(o) && DateTimeType.class.isAssignableFrom(o)))) {
+            dataType = "datetime";
         } else {
-            dataType = "Text";
+            dataType = "text";
         }
         snippet = snippet.replace("%data_type%", dataType);
 
+        String inputHint = input.getInputHint();
+        snippet = snippet.replace("%input_hint%", inputHint == null ? "" : inputHint);
+
+        String inputType = "text";
+        String inputPattern = "";
+        if ("number".equals(inputHint)) {
+            inputType = "number";
+            inputPattern = "pattern=\"[0-9]*([\\.|,][0-9]*)?\"";
+        } else if ("date".equals(inputHint)) {
+            inputType = "date";
+            inputPattern = "pattern=\"[0-9]{4}-[0-9]d{2}-[0-9]d{2}\"";
+        } else if ("time".equals(inputHint)) {
+            inputType = "time";
+            inputPattern = "pattern=\"[0-9]{2}:[0-9]{2}(:[0-9]{2})?\"";
+        } else if ("datetime".equals(inputHint)) {
+            inputType = "datetime-local";
+            inputPattern = "pattern=\"[0-9]{4}-[0-9]d{2}-[0-9]d{2} [0-9]{2}:[0-9]{2}(:[0-9]{2})?\"";
+        }
+        snippet = snippet.replace("%input_type%", inputType);
+        snippet = snippet.replace("%input_pattern%", inputPattern);
+
+        String displayState = getValue(w);
+        State state = itemUIRegistry.getState(w);
+
+        String undefState = "-";
+        if ("datetime".equals(dataType)) {
+            undefState = "";
+            if ((inputHint == null || inputHint.isEmpty())) {
+                undefState = "YYYY-MM-DD hh:mm:ss";
+            }
+        } else if (state == null || state instanceof UnDefType) {
+            undefState = displayState;
+            if ("number".equals(inputHint)) {
+                String[] stateArray = displayState.trim().split(" ");
+                if (stateArray.length > 0) {
+                    undefState = stateArray[0];
+                }
+            }
+        }
+        snippet = snippet.replace("%undef_state%", undefState);
+
+        String dataState = displayState;
```

Comment:
No need to assign a value as you assign it in the if and in the else.
Or you could assign it to "" and then avoid an else in your if.